### DISCUSSION TITLE
Randomness in pyMOR

### DISCUSSION THREAD

**Discussion by sdrave** [pymor/pymor#1577] [Ideas]
Currently, randomization in pyMOR is not handled in a uniform way. Some methods have `random_state` and `seed` parameters which they use to call `pymor.tools.random.get_random_state`. On the other hand, `samdp` has a `seed` default which is used to seed numpy's default random generator.

From #1564:

> I quite like the idea of removing the `seed` parameter everywhere and replacing it by `random_state` (where it is missing). That would mean that the methods in `lrradi`, `samdp`, `eigs` would loose their `seed` `defaults` replacing them by a global `default` `seed` for the default random state. Thus, as already said above, this would mean that consecutive calls to these methods would lead to different results, deterministically determined by the global seed and the execution order. If one would want to have the same random state one would have to pass a manually seeded random state to the method. I would assume that this is rarely needed. What do you think @pmli, @lbalicki?

Up to randomization that is difficult or expensive to avoid (e.g., parallelization in BLAS), I think having as much determinism as possible is desirable (e.g., to have more-or-less reproducible code for publications). So, although it might seem unintuitive, I would be for methods, that use random numbers, to return the same result on consecutive calls. I was actually thinking that we should discuss [JAX's approach](https://jax.readthedocs.io/en/latest/jax-101/05-random-numbers.html) (basically, making `seed` a required argument without a default value, and a way to generate new seeds) and whether it makes sense to do something similar in pyMOR.

**Comment by sdrave** [comment]
Some thoughts:
- I agree that results should be reproducible in the sense that two runs of the same program with the same data should give the same result. This is (almost, see below) the case for the `random_state` proposal.
- Always using the same seed for an algorithm which involves randomness is actually bad because it makes you blind for the influence of the randomness on the result. A typical case where this happens is when you pick a too small random validation set to compute some error.
- JAX's approach would force us to add seed arguments to every method that might (indirectly) call a method which depends on randomness.
- JAX only does this because determinism is no longer guaranteed when randomized code is executed in parallel. However, the only case where this could become an issue in pyMOR is through a `WorkerPool`. It should be quite easy to ensure that each job for a `WorkerPool` gets a new initial random state which deterministically depends on the state when the job is queued.
- It is currently unclear how pyMOR will develop in the future regarding parallelization. However, I do not see pyMOR ever becoming fully thread safe. For any task-scheduling system it should be possible to have deterministically initialized task-local random states. However, one would run into problems for code that executes randomized code in an order depending on the non-deterministic order in which child tasks finish. I can't really imagine many cases where this might happen, and these cases could the be dealt with.
- Overall this makes me think, whether we should actually remove all `seed` and `random_state` parameters and go with a `random_state` context manager which temporarily sets a new (task-local) global random state.

**Reply by artpelling** [reply]
Some additional thoughts:

What kind of behaviour do we want to achieve? I understand that "running a program" means to execute some python script via the interpreter and "calling a method" refers to a specific line of that code. Disregarding the problems of parallelization, it should sufficient (for reproducible publications and so on) if multiple runs of the program give the same result.
 
Wouldn't a mandatory seed argument (the JAX approach) lead to problems?
There are multiple layers of randomized methods in pyMOR, for example `randomized_gerneralized_svd` calls `rrf` which calls `VectorArray.random` which calls `VectorSpace.random`. Maybe I got something wrong, but the only sensible use of a mandatory seed argument I can see is with the "top-level" method `randomized_generalized_svd` such that it returns the same results on multiple calls. However, IMO all of these randomized methods should be treated equally, because all of them can be "on top". If `VectorSpace.random` had a mandatory seed argument and multiple calls returned the same values, all random vectors would have to be created in one line of the code and iterative randomization loops like in `adaptive_rrf` wouldn't be correct anymore.

Again, maybe I missed something or maybe we all actually mean the same thing, but I also think the `random_state` proposal is the way to go. (I can't say much about the parallelization issue).

**Reply by renefritze** [reply]
> * Overall this makes me think, whether we should actually remove all `seed` and `random_state` parameters and go with a `random_state` context manager which temporarily sets a new (task-local) global random state

This is effectively what we do in [some tests](https://github.com/pymor/pymor/blob/ac8bed091acd732506d5a32fae6340dbbe8eaf4d/src/pymortests/vectorarray.py#L658) already, via [Hypothesis](https://hypothesis.readthedocs.io/en/latest/data.html#hypothesis.strategies.random_module)

I quite like this approach. Ofc this can only ever manage known sources of randomness, but I don't think that's a real-world downside. Should be noted in a docstring for such a decorator maybe. IIRC hypothesis also has a mechanism for other PRNGs to hook into the state management.

**Reply by pmli** [reply]
> - Overall this makes me think, whether we should actually remove all `seed` and `random_state` parameters and go with a `random_state` context manager which temporarily sets a new (task-local) global random state.

I don't really understand this proposal. Does it mean that, to have a reproducible script, I need to indent everything and add something like `with random_state(...):` at the top? That doesn't sound so nice, especially if I also want to use it in a Jupyter notebook.

**Comment by pmli** [comment]
I spent some time trying to understand PRNG in NumPy. Some main points I got were:

- NumPy, since version 1.17, recommends using `Generators` instead of `RandomState` (see [here](https://numpy.org/doc/stable/reference/random/index.html))
- NumPy recommends using `SeedSequence` to generate seeds (see [here](https://numpy.org/doc/stable/reference/random/bit_generators/index.html#seeding-and-entropy) and [here](https://numpy.org/doc/stable/reference/random/parallel.html#seedsequence-spawning))

Although the discussion started on the topic of `random_state` and `seed` parameters, we might need to discuss:

1. replacing calls to `np.random.RandomState` by `np.random.default_rng` and
2. replacing `np.random.seed` with a combination of `np.random.SeedSequence` and `np.random.default_rng`

**Comment by pmli** [comment]
Looking at NumPy's documentation and also at this [issue](https://github.com/numpy/numpy/issues/15322), I guess there are a few possible approaches to achieve what I would like (determinism by default):

1. all random functions accept a `Generator` called `rg`
    1. if a random function `f` calls another random function `g`, `f` passes `rg` to `g`
    2. if a random function wants to generate random numbers in parallel, it uses `rg.bit_generator._seed_seq.spawn` to generate new `SeedSequences` which can be turned into `Generators` using `np.random.default_rng`
    3. if a random function needs to directly generate random numbers, it uses `rg`
2. all random functions accept a `SeedGenerator` called `ss`
    1. if a random function `f` calls another random function `g`, `f` passes `ss.spawn(1)[0]` to `g`
    2. if a random function wants to generate random numbers in parallel, it uses `ss.spawn` to generate new `SeedSequences`
    3. if a random function needs to directly generate random numbers, it uses `np.random.default_rng(ss)` to build a `Generator`
3. some combination of the above

There could be an option to pass `None`, which both `np.random.default_rng` and `np.random.SeedSequence` interpret as "use OS entropy", and so could be used to have "truly" random numbers, but I personally don't find it desirable.

The users have the options:

1. build a `Generator` and pass it to every random function,
2. build a `SeedSequence` and use its `spawn` method to generate new `SeedSequences` that can be passed to random functions.
